\input{config}

\title{Theoretical and practical metagenomic approaches to viral discovery}
\subtitle{Practical Session: Random Forest Classifier and viral application}
\author{Kevin Lamkiewicz, Manja Marz}
\date{23.10.2019\\[1em]European Virus Bioinformatics Center}

\begin{document}

\begin{frame}
  \maketitle
\end{frame}

\section[Implementation]{How to: Implement the Random Forest Classifier}

\begin{frame}[c,fragile]\frametitle{Random Forest}
  \begin{lstlisting}[language=Python]
from sklearn.ensemble import RandomForestClassifier
from sklearn import datasets
# we know this already...
iris = datasets.load_iris()
data = iris.data
target = iris.target
@\pause@

# create the classifier object
# n_estimators: number of trees for the forest
# max_depth: maximum depth of one tree
rfc = RandomForestClassifier(n_estimators=100, max_depth=2)
rfc = rfc.fit(data, target)
  \end{lstlisting}
\end{frame}


\section[Hands-On]{Training our model to identify viral elements}

\begin{frame}[c]\frametitle{We want to identify viral pre-miRNAs}
  \begin{block}{Our task:}
    In the \texttt{miRBase} database are currently around 320 viral pre-miRNAs.
    We want to use them in to train a machine learning model that can distinguish between
    viral pre-miRNAs and other sequences.
  \end{block}
  \vspace*{3em}
  \uncover<2->{
    \begin{block}{I already prepared...}
      \begin{itemize}
        \item a file with all precursors
        \item a file with our negative data
        \item some utilities we need (reading the data, ...)
      \end{itemize}
    \end{block}
  }
\end{frame}

\begin{frame}[c,fragile]\frametitle{}
  \begin{lstlisting}[language=Python]
# first we need to translate the sequences into something
# that can be used by the machine learning algorithm:
# what are possible features for our task?
@\pause@
import numpy as np
@\pause@
# read in the data
trainingsData = read_training_set(positive, negative)
@\pause@
# transform the sequence into features
trainingSet, targets = transform_data(trainingsData)
  \end{lstlisting}
\end{frame}

% Backup Slides. Using this macro, you'd get a slide number for each
% backup slide without increasing the maximum slide numbers of the original presentation.
% However, for this, the framenumber has to be inserted - which isn't in the template by default
\beginbackup

\begin{frame}[c]\frametitle{Coffee Break}
  \begin{figure}[htbp]
    \centering
    \includegraphics[width=0.65\textwidth]{coffeebreak.png}
  \end{figure}
\end{frame}

\backupend

\end{document}