\input{config}

\title{Theoretical and practical metagenomic approaches to viral discovery}
\subtitle{Practical Session: LRIscan for viral long-range RNA-RNA Interactions}
\author{Kevin Lamkiewicz, Manja Marz}
\date{25.10.2019\\[1em]European Virus Bioinformatics Center}

\begin{document}

\begin{frame}
  \maketitle
\end{frame}

\section[Alignment Recap]{Alignments and compensatory mutations}

\begin{frame}[c]\frametitle{Unconserved sequence, conserved structure}
	\includegraphics[width=\textwidth]{figures/color_code.pdf}
\end{frame}


\begin{frame}[c]\frametitle{Unconserved sequence, conserved structure}
	\includegraphics[width=\textwidth]{figures/cov_color_aln.ps}
\end{frame}

\begin{frame}[c]\frametitle{Compensatory Mutations in secondary structures}
	\begin{block}{Importance of such mutations}
		Compensatory mutations underline the importance of a specific secondary structure.
	\end{block}

	\uncover<2->{
	\begin{block}{But be careful!}
		If we're assuming a uniform mutation rate, every third pair of mutations is a compensatory mutation.
	\end{block}
	\small
	\begin{minipage}{0.4\textwidth}
	\centering
	\begin{tabular}{cc}
		\toprule
		A & U \\
		\midrule
		A & A \\
		A & C \\
		A & G \\
		C & A \\
		C & C \\
		\textcolor{blue}{C} & \textcolor{blue}{G} \\
		C & U \\
		\bottomrule
	\end{tabular}
	\end{minipage} \hfill \begin{minipage}{0.4\textwidth}
	\centering
	\begin{tabular}{cc}
		\toprule
		A & U \\
		\midrule
		G & A \\
		\textcolor{blue}{G} & \textcolor{blue}{C} \\
		G & G \\
		\textcolor{blue}{G} & \textcolor{blue}{U} \\
		\textcolor{blue}{U} & \textcolor{blue}{A} \\
		U & C \\
		\textcolor{blue}{U} & \textcolor{blue}{G} \\
		U & U \\
		\bottomrule
	\end{tabular}
	\end{minipage}
	}
\end{frame}

\section[LRIs]{RNA-RNA Long-Range Interactions}

\begin{frame}[c]\frametitle{Why LRIs?}
 \begin{itemize}
 	\item Interaction spans distances between a few hundred and several thousands of nucleotides
 	\item few are described in positive stranded RNA viruses
 	\item often located in loop regions (bulges, hairpins, ...)\\
 	\uncover<2->{
 	$\Rightarrow$ pseudo-knots!
 	} \uncover<3->{
 	\item LRIs may play a very important role in viral replication
 	}
 \end{itemize}
\end{frame}

\begin{frame}[c]\frametitle{How to calculate LRIs}

	\begin{minipage}[t]{0.3\textwidth}
	\begin{block}{Approach I}
	\begin{itemize}
		\item RNAduplex
		\item RNAplex
		\item RNAhybrid
	\end{itemize}
	\end{block}
	\uncover<4->{
	\begin{block}{Approach IV}
		\begin{itemize}
			\item inteRNA
			\item inRNAs
		\end{itemize}
	\end{block}
	}
	\end{minipage}
	\begin{minipage}[t]{0.3\textwidth}
	\uncover<2->{
	\begin{block}{Approach II}
	\begin{itemize}
		\item RNAcofold
	\end{itemize}
	\end{block}
	}
	\vspace*{3.5em}
	\uncover<5->{
	\begin{block}{Approach V}
		\begin{itemize}
			\item PETcofold
			\item RNAaliduplex
		\end{itemize}
	\end{block}
	}
	\end{minipage}
	\begin{minipage}[t]{0.3\textwidth}
	\uncover<3->{
	\begin{block}{Approach III}
		\begin{itemize}
			\item RNAup
			\item IntaRNA
		\end{itemize}
	\end{block}
	}
	\end{minipage}
\end{frame}

\begin{frame}[c]\frametitle{LRIscan}
	\textbf{Prediction of conserved long-range RNA-RNA interactions in full viral genomes}, 2016. M.~Fricke, M.~Marz \\
	\uncover<2->{
	$\Rightarrow$ \texttt{LRIscan}
	}
\end{frame}

\section[LRIscan]{How does LRIscan work and how do I use it?}

\begin{frame}[c]\frametitle{Workflow of LRIscan}
	\begin{center}
		\includegraphics[height=0.9\textheight]{figures/lri_workflow_1.pdf}
	\end{center}
\end{frame}

\begin{frame}[c]\frametitle{Coverage and Complexity}
	\begin{minipage}[t]{0.45\textwidth}
		\begin{block}{Coverage of an alignment}
			Relative number of sequences that do not have a gap on a specific position.
		\end{block}
	\end{minipage} \hfill \uncover<2->{
	\begin{minipage}[t]{0.45\textwidth}
		\begin{block}{Complexity of the alignment}
			\[
				C_i = \frac{1}{m} \sum_{k=1}^{m} \frac{\lvert \delta(a_{i\dots i+s-1}^k)\rvert}{\lvert (a_{i\dots i+s-1}^k) \rvert}
			\]

			\uncover<3->{
			\[
				\delta(CCUUUGGAAA) = CUGA
			\]
			}

		\end{block}
	\end{minipage}
	}
\end{frame}

\begin{frame}[c]\frametitle{Workflow of LRIscan -- Step 2}
	\begin{center}
		\includegraphics[height=0.9\textheight]{figures/lri_workflow_2.pdf}
	\end{center}
\end{frame}

\begin{frame}[c]\frametitle{Finding seeds}
	\[
		S_{i,j} = (S_{i-1,j+1} + 1) \cdot \Pi_{ij} \cdot \Phi_{ij}
	\]
	\begin{itemize}
		\item $\Pi_{ij}$: do at least $t$ percent of the input sequence form the basepair $(i,j)$?
		\item $\Phi_{ij}:$ do both alignment columns $A_i$ and $A_j$ meet the coverage threshold?
	\end{itemize}
\end{frame}


\begin{frame}[c]\frametitle{Workflow of LRIscan -- Step 3}
	\begin{center}
		\includegraphics[height=0.9\textheight]{figures/lri_workflow_3.pdf}
	\end{center}
\end{frame}

\begin{frame}[c]\frametitle{Seed Scoring}
	\begin{itemize}
		\item z-Score analysis for each seed to measure reliability
		\item compensatory score $\tau$
	\end{itemize}
	\[
		\tau = \frac{\sum_b(u\cdot h)}{6\cdot \lvert b\rvert\cdot k}
	\]
	with:
	\begin{itemize}
		\item $u$: number of different base-pair types
		\item $h$: number of incompatible base-pairs 
	\end{itemize}
\end{frame}

\begin{frame}[c]\frametitle{Workflow of LRIscan -- Step 4}
	\begin{center}
		\includegraphics[height=0.9\textheight]{figures/lri_workflow_final.pdf}
	\end{center}
\end{frame}

\begin{frame}[c]\frametitle{Seed Extension}
	\begin{itemize}
		\item each seed is extended 10\,nts at the 5' (and 3' respectively)
		\item calculate MFE with RNAalifold
		\begin{itemize}
			\item hard constraints for seed region
			\item soft constraints for extension, such that intermolecular interactions are formed
		\end{itemize}
	\end{itemize}
\end{frame}

\section[Results]{LRIscan Results and Output}

\begin{frame}[c, fragile]\frametitle{LRIscan Usage}
	\begin{lstlisting}
$> ./LRIscan.rb -c 2 -f <ALIGNMENT> -o <OUTPUT>
	\end{lstlisting}

\begin{itemize}
	\item tabular output in .tsv format
	\item table and figures in .html
	\item all figures are also stored in the \texttt{ps/} directory
\end{itemize}
\end{frame}

\begin{frame}[c,fragile]{LRIscan Hands-On}
  \begin{block}{Exercise:}
    Go to \url{https://www.rna.uni-jena.de/supplements/lriscan/}\\
    \begin{enumerate}
      \item Download the MSA of the Flaviviruses
      \item Apply \texttt{LRIscan}
      \item Do not look at the results on the webpage (yet)
    \end{enumerate}
  \end{block}
\end{frame}

\section[circos]{Visualisation of LRIs with circos}

\begin{frame}[c]\frametitle{What is circos in the first place?}
	\centering
	\url{http://circos.ca/}
\end{frame}

\begin{frame}[c]\frametitle{What we'll do:}
	\begin{itemize}
		\item create a circular genome of DENV
		\item extract the potential LRIs from the LRIscan results
		\item visualize them using circos \\[2em]
		\item<2> being happy! 
	\end{itemize}
\end{frame}

\begin{frame}[c]\frametitle{Karyo}
\end{frame}

% Backup Slides. Using this macro, you'd get a slide number for each
% backup slide without increasing the maximum slide numbers of the original presentation.
% However, for this, the framenumber has to be inserted - which isn't in the template by default
\beginbackup

\begin{frame}[c]\frametitle{Coffee Break}
  \begin{figure}[htbp]
    \centering
    \includegraphics[width=0.65\textwidth]{coffeebreak.png}
  \end{figure}
\end{frame}

\backupend

\end{document}